### DISCUSSION TITLE
mongoose connect waiting for create collection

### DISCUSSION THREAD

**Discussion by kathirsvn** [stargate/stargate-mongoose#33] [General]
Hi @vkarpov15 
While running a test (called 'StargateMongoose - index') in [tests/driver/index.test.ts](https://github.com/stargate/stargate-mongoose/blob/main/tests/driver/index.test.ts), I've got a couple of questions (1. with respect to an option for the connect function 2. the 'await' for connect to complete). So I would like to understand more about this part of the mongoose library.
### Code
```typescript
const productSchema = new mongoose.Schema({
  name: String,
  price: Number
});
const Product = mongoose.model('Product', productSchema);
it('should leverage astradb', async function () {    
    await mongoose.connect(astraUri, {
      username: "cassandra",
      password: "cassandra",
      authUrl: "http://localhost:8081/v1/auth"
    });
    //await new Promise(fn => setTimeout(fn, 3000));//TODOV3 check later - without this delay, this test fails sometimes
    const product1 = new Product({ name: 'Product 1', price: 10 });
    await product1.save();

    const product2 = new Product({ name: 'Product 2', price: 10 });
    await product2.save();
});
```

Before trying to save any document in the `products` collection, the idea is that the 'createCollection' must be invoked for products' while connecting to the backend.

### Question 1: 
The connect function has a set of options and one of them is 'autoCreate' which is set to `false` by default as per [the doc](https://mongoosejs.com/docs/api/mongoose.html#mongoose_Mongoose-connect), which takes care of creating collection automatically.  Though, we don't set it to `true`, explicitly, how the `createCollection` is called while invoking the `connect` function in the stargate-mongoose driver workflow?

### Question 2:
Without the below delay code, which I commented in the code above, the test **_sometimes fails_** with table doesn't exist error
```typescript
await new Promise(fn => setTimeout(fn, 3000));
```
```log
error: Error: Command "insertOne" failed with the following errors: [{"message":"INVALID_ARGUMENT: table testks1.products does not exist","exceptionClass":"StatusRuntimeException"}]
    at handleIfErrorResponse (/Users/kathiresan.selvaraj/Documents/workspace/stargate-mongoose/src/client/httpClient.ts:15:25)
    at HTTPClient.executeCommand (/Users/kathiresan.selvaraj/Documents/workspace/stargate-mongoose/src/client/httpClient.ts:15:25)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at /Users/kathiresan.selvaraj/Documents/workspace/stargate-mongoose/src/collections/collection.ts:25:259
    at executeOperation (/Users/kathiresan.selvaraj/Documents/workspace/stargate-mongoose/src/collections/utils.ts:55:19)
    1) should leverage astradb
```
So, why the `await mongoose.connect` call doesn't wait for the `createCollection` call to be completed?

cc: @jeffreyscarpenter @amorton

**Comment by vkarpov15** [comment]
@kathirsvn I think for stargate-mongoose, we should ask users to set `autoCreate` to `true`. Mongoose doesn't set `autoCreate` by default because `autoCreate` usually isn't useful for MongoDB - in most cases, an operation on a non-existent collection results in MongoDB creating the collection for you. There are exceptions, most notably multi-document transactions; writing to a non-existent collection in a transaction throws an error.

`createCollection()` is not called while invoking `connect()`. Mongoose calls `createCollection()` in [`Model.init()`](https://github.com/Automattic/mongoose/blob/master/lib/model.js#L1301-L1317), which Mongoose calls when creating a model using `mongoose.model()` or `Connection.prototype.model()`. Stargate-mongoose should **not** call `createCollection()` in `connect()`, because the user can create models after `connect()` is done.

For the `testks1.products does not exist`, adding `autoCreate: true` and calling `await Model.init()` to wait for `createCollection()` to finish should be enough to fix that error. For example, [in this file in the sample app](https://github.com/stargate/stargate-mongoose-sample-apps/pull/6/files#diff-d0adce0e360a16ca06ddb9b26c7db6e2ca9ce467ff79b200ae937cad2f35c20e) we set `autoCreate` and then wait for all models to finish `init()`.

Does that help?

**Reply by kathirsvn** [reply]
@vkarpov15 
Thanks a lot for the details. This really helps. I had a misconception that `connect()` invokes `createCollection()` when the flag `autoCreate` is set to `true` and I'm clear on that part now.

**A follow up question on the `autoCreate` flag in the sample apps code shared:** 

When `await Model.init()` is going to call `createCollection()` and wait for it to complete, what is the usage of `autoCreate: true` here?

**Reply by jeffreyscarpenter** [reply]
Can we default `autoCreate: true` for stargate-mongoose to make this simpler for users?

**Reply by vkarpov15** [reply]
`await Model.init()` doesn't call `createCollection()`, or at least not if the user calls it. `mongoose.model()` calls `Model.init()` under the hood, and `Model.init()` is idempotent. So user calling `await Model.init()` means "wait for `createCollection()` to finish".

Admittedly, this API is a bit quirky. But it works reasonably well for MongoDB.

Mongoose doesn't have a way for drivers to set default options currently. Mongoose can add support for that, but best bet for now is to ask users to set `autoCreate` themselves
